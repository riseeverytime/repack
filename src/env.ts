export const WORKER_ENV_KEY = 'RNWT_WORKER';

export const VERBOSE_ENV_KEY = 'RNWT_VERBOSE';

export const CLI_OPTIONS_ENV_KEY = 'RNWT_CLI_OPTIONS';

/**
 * Checks if code is running as a worker.
 *
 * @returns True if running as a worker.
 *
 * @internal
 */
export function isWorker() {
  return Boolean(process.env[WORKER_ENV_KEY]);
}

/**
 * Checks if code is running in verbose mode.
 *
 * @returns True if running in verbose mode.
 *
 * @internal
 */
export function isVerbose() {
  return Boolean(process.env[VERBOSE_ENV_KEY]);
}
