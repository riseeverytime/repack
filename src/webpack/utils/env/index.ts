export * from './getContext';
export * from './getReactNativePath';
export * from './getPlatform';
export * from './getMode';
export * from './getEntry';
export * from './isMinimizeEnabled';
export * from './getDevServerOptions';
