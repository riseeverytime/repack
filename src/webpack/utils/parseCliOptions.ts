import path from 'path';
import { CliOptions, WebpackOptions } from '../../types';

export const CLI_OPTIONS_KEY = 'HAUL2_CLI_OPTIONS';

export type WebpackOptionsWithoutPlatform = Omit<WebpackOptions, 'platform'>;

export type OptionalWebpackOptions = {
  [K in keyof WebpackOptionsWithoutPlatform]?: WebpackOptions[K];
};

export interface FallbackWebpackOptions extends OptionalWebpackOptions {
  platform: WebpackOptions['platform'];
}

export interface ParseCliOptionsConfig {
  fallback: FallbackWebpackOptions;
}

export const DEFAULT_FALLBACK: WebpackOptionsWithoutPlatform = {
  mode: 'development',
  dev: true,
  entry: './index.js',
  outputPath: path.join(process.cwd(), 'dist'),
  assetsOutputPath: path.join(process.cwd(), 'dist'),
  outputFilename: 'index.bundle',
  context: process.cwd(),
  reactNativePath: path.join(process.cwd(), './node_modules/react-native'),
  minimize: false,
};

export const DEFAULT_PORT = 8081;

export function parseCliOptions(config: ParseCliOptionsConfig): WebpackOptions {
  const fallback: WebpackOptions = { ...DEFAULT_FALLBACK, ...config.fallback };
  const rawCliOptions = process.env[CLI_OPTIONS_KEY];
  if (!rawCliOptions) {
    return fallback;
  }

  const cliOptions: CliOptions = JSON.parse(rawCliOptions);

  if ('bundle' in cliOptions.arguments) {
    const args = cliOptions.arguments.bundle;

    const outputPath = path.dirname(args.bundleOutput);
    const outputFilename = path.basename(args.bundleOutput);
    const entry = args.entryFile;

    return {
      mode: args.dev ? 'development' : 'production',
      dev: args.dev,
      context: cliOptions.config.root,
      platform: args.platform,
      entry: entry.startsWith('./') ? entry : `./${entry}`,
      outputPath,
      outputFilename,
      sourcemapFilename: args.sourcemapOutput,
      assetsOutputPath: args.assetsDest,
      minimize: Boolean(args.minify),
      reactNativePath: cliOptions.config.reactNativePath,
    };
  } else if ('start' in cliOptions.arguments) {
    const args = cliOptions.arguments.start;

    return {
      mode: 'development',
      dev: true,
      context: cliOptions.config.root,
      platform: args.platform,
      entry: fallback.entry,
      outputPath: fallback.outputPath,
      outputFilename: `index.${args.platform}.bundle`,
      sourcemapFilename: undefined,
      assetsOutputPath: undefined,
      minimize: false,
      reactNativePath: cliOptions.config.reactNativePath,
      devServer: {
        host: args.host || undefined,
        port:
          args.port ??
          (typeof fallback.devServer === 'boolean' ||
          fallback.devServer?.port === undefined
            ? DEFAULT_PORT
            : fallback.devServer.port),
        https: args.https,
        cert: args.cert || undefined,
        key: args.key || undefined,
      },
    };
  }

  throw new Error('todo: should never happen');
}
