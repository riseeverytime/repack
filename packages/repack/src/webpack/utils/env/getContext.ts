import { Fallback } from '../../../types';
import { getFallbackFromOptions } from './internal/getFallbackFromOptions';
import { parseCliOptions } from './internal/parseCliOptions';

export function getContext(
  options: Fallback<string> = { fallback: process.cwd() }
): string {
  const cliOptions = parseCliOptions();
  if (!cliOptions) {
    return getFallbackFromOptions(options);
  }

  return cliOptions.config.root;
}
