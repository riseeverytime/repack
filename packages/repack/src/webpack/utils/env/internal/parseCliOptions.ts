import { CLI_OPTIONS_ENV_KEY } from '../../../../env';
import { CliOptions } from '../../../../types';

export function parseCliOptions(): CliOptions | undefined {
  const rawCliOptions = process.env[CLI_OPTIONS_ENV_KEY];
  if (!rawCliOptions) {
    return undefined;
  }
  return JSON.parse(rawCliOptions);
}
