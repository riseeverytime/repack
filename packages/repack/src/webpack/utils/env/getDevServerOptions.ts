import type { DevServerOptions, Fallback } from '../../../types';
import { getFallbackFromOptions } from './internal/getFallbackFromOptions';
import { parseCliOptions } from './internal/parseCliOptions';

/** Default development server (proxy) port. */
export const DEFAULT_PORT = 8081;

type DeepOptional<T> = T extends { [key: string]: any }
  ? {
      [K in keyof T]?: DeepOptional<T[K]>;
    }
  : T | undefined;

export function getDevServerOptions(
  options: Fallback<DeepOptional<DevServerOptions>> = {
    fallback: { port: DEFAULT_PORT },
  }
): DevServerOptions {
  const cliOptions = parseCliOptions();
  if (!cliOptions) {
    return {
      port: DEFAULT_PORT,
      enabled: false,
      ...getFallbackFromOptions(options),
    };
  }

  if ('bundle' in cliOptions.arguments) {
    return {
      port: DEFAULT_PORT,
      enabled: false,
      ...getFallbackFromOptions(options),
    };
  } else {
    const { host, port, https, cert, key } = cliOptions.arguments.start;
    return {
      enabled: true,
      hmr: getFallbackFromOptions(options).hmr ?? true,
      host: host || getFallbackFromOptions(options).host,
      port: port ?? getFallbackFromOptions(options).port ?? DEFAULT_PORT,
      https: https ? { cert, key } : undefined,
    };
  }
}
