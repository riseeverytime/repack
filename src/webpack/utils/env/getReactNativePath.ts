import { Fallback } from '../../../types';
import { getFallbackFromOptions } from './internal/getFallbackFromOptions';
import { parseCliOptions } from './internal/parseCliOptions';

export function getReactNativePath(
  options: Fallback<string> = { fallback: require.resolve('react-native') }
): string {
  const cliOptions = parseCliOptions();
  if (!cliOptions) {
    return getFallbackFromOptions(options);
  }

  return cliOptions.config.reactNativePath;
}
