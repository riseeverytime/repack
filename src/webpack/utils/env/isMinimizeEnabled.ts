import { Fallback } from '../../../types';
import { getFallbackFromOptions } from './internal/getFallbackFromOptions';
import { parseCliOptions } from './internal/parseCliOptions';

export function isMinimizeEnabled(
  options: Fallback<boolean> = { fallback: false }
): boolean {
  const cliOptions = parseCliOptions();
  if (!cliOptions) {
    return getFallbackFromOptions(options);
  }

  if ('bundle' in cliOptions.arguments) {
    return (
      cliOptions.arguments.bundle.minify ?? !cliOptions.arguments.bundle.dev
    );
  } else {
    return getFallbackFromOptions(options);
  }
}
