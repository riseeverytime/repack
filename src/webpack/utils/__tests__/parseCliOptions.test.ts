import { CLI_OPTIONS_ENV_KEY } from '../../../env';
import { parseCliOptions } from '../parseCliOptions';

afterEach(() => {
  delete process.env[CLI_OPTIONS_ENV_KEY];
});

describe('parseCliOptions', () => {
  describe('for creating static bundle', () => {
    it('should return fallback values if CLI options are empty', () => {
      const options = parseCliOptions({ fallback: { platform: 'xyz' } });
      expect(options).toEqual({
        context: '/a/b/c',
        dev: true,
        mode: 'development',
        minimize: false,
        entry: './index.js',
        outputFilename: 'index.bundle',
        outputPath: '/a/b/c/dist',
        platform: 'xyz',
        reactNativePath: '/a/b/c/node_modules/react-native',
        sourcemapFilename: '[file].map',
      });
    });

    it('should return values based on CLI options - fixture 1', () => {
      process.env[CLI_OPTIONS_ENV_KEY] = JSON.stringify({
        config: {
          root: '/a/b/c',
          reactNativePath: '/a/b/c/node_modules/react-native',
          webpackConfigPath: '/a/b/c/webpack.config.js',
        },
        command: 'bundle',
        arguments: {
          bundle: {
            bundleOutput: '/x/y/z/main.bundle',
            assetsDest: '/x/y/z/app',
            entryFile: '/a/b/c/index.js',
            sourcemapOutput: '/x/y/z/main.bundle.map',
            dev: false,
            platform: 'xyz',
            minify: true,
          },
        },
      });
      const options = parseCliOptions({ fallback: { platform: 'xyz' } });
      expect(options).toEqual({
        context: '/a/b/c',
        dev: false,
        mode: 'production',
        minimize: true,
        entry: '/a/b/c/index.js',
        outputFilename: '../main.bundle',
        outputPath: '/x/y/z/app',
        platform: 'xyz',
        reactNativePath: '/a/b/c/node_modules/react-native',
        sourcemapFilename: '../main.bundle.map',
      });
    });

    describe('on iOS platform', () => {
      it('should return fallback values if CLI options are empty', () => {
        const options = parseCliOptions({ fallback: { platform: 'ios' } });
        expect(options).toMatchSnapshot();
      });

      it('should return values based on CLI options - fixture 1', () => {
        process.env[CLI_OPTIONS_ENV_KEY] = JSON.stringify({
          config: {
            root: '/a/b/c',
            reactNativePath: '/a/b/c/node_modules/react-native',
            webpackConfigPath: '/a/b/c/webpack.config.js',
          },
          command: 'bundle',
          arguments: {
            bundle: {
              bundleOutput:
                '/x/y/z/DerivedData/TesterApp-xxxxxx/Build/Products/Release-iphonesimulator/main.jsbundle',
              assetsDest:
                '/x/y/z/DerivedData/TesterApp-xxxxxx/Build/Products/Release-iphonesimulator/TesterApp.app',
              entryFile: 'index.js',
              sourcemapOutput: undefined,
              dev: false,
              platform: 'ios',
              minify: true,
            },
          },
        });
        const options = parseCliOptions({ fallback: { platform: 'ios' } });
        expect(options).toMatchSnapshot();
      });
    });

    describe('on Android platform', () => {
      it('should return fallback values if CLI options are empty', () => {
        const options = parseCliOptions({ fallback: { platform: 'android' } });
        expect(options).toMatchSnapshot();
      });

      it('should return values based on CLI options - fixture 1', () => {
        process.env[CLI_OPTIONS_ENV_KEY] = JSON.stringify({
          config: {
            root: '/a/b/c',
            reactNativePath: '/a/b/c/node_modules/react-native',
            webpackConfigPath: '/a/b/c/webpack.config.js',
          },
          command: 'bundle',
          arguments: {
            bundle: {
              bundleOutput:
                '/x/y/z/build/generated/assets/react/release/index.android.bundle',
              assetsDest: '/x/y/z/build/generated/res/react/release',
              entryFile: 'index.js',
              sourcemapOutput:
                '/x/y/z/build/generated/sourcemaps/react/release/index.android.bundle.map',
              dev: false,
              platform: 'android',
              minify: true,
            },
          },
        });
        const options = parseCliOptions({ fallback: { platform: 'android' } });
        expect(options).toMatchSnapshot();
      });
    });
  });
});
