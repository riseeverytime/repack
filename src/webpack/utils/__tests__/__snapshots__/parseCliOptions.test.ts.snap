// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCliOptions for creating static bundle on Android platform should return fallback values if CLI options are empty 1`] = `
Object {
  "context": "/a/b/c",
  "dev": true,
  "entry": "./index.js",
  "minimize": false,
  "mode": "development",
  "outputFilename": "index.bundle",
  "outputPath": "/a/b/c/dist",
  "platform": "android",
  "reactNativePath": "/a/b/c/node_modules/react-native",
  "sourcemapFilename": "[file].map",
}
`;

exports[`parseCliOptions for creating static bundle on Android platform should return values based on CLI options - fixture 1 1`] = `
Object {
  "context": "/a/b/c",
  "dev": false,
  "entry": "./index.js",
  "minimize": true,
  "mode": "production",
  "outputFilename": "../../../assets/react/release/index.android.bundle",
  "outputPath": "/x/y/z/build/generated/res/react/release",
  "platform": "android",
  "reactNativePath": "/a/b/c/node_modules/react-native",
  "sourcemapFilename": "../../../sourcemaps/react/release/index.android.bundle.map",
}
`;

exports[`parseCliOptions for creating static bundle on iOS platform should return fallback values if CLI options are empty 1`] = `
Object {
  "context": "/a/b/c",
  "dev": true,
  "entry": "./index.js",
  "minimize": false,
  "mode": "development",
  "outputFilename": "index.bundle",
  "outputPath": "/a/b/c/dist",
  "platform": "ios",
  "reactNativePath": "/a/b/c/node_modules/react-native",
  "sourcemapFilename": "[file].map",
}
`;

exports[`parseCliOptions for creating static bundle on iOS platform should return values based on CLI options - fixture 1 1`] = `
Object {
  "context": "/a/b/c",
  "dev": false,
  "entry": "./index.js",
  "minimize": true,
  "mode": "production",
  "outputFilename": "../main.jsbundle",
  "outputPath": "/x/y/z/DerivedData/TesterApp-xxxxxx/Build/Products/Release-iphonesimulator/TesterApp.app",
  "platform": "ios",
  "reactNativePath": "/a/b/c/node_modules/react-native",
  "sourcemapFilename": "[file].map",
}
`;
